internal/provider/resource_gitlab_project_runner_enablement.go

```diff
@@ -0,0 +1,144 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+				Description: "The ID or URL-encoded path of the project owned by the authenticated user",
+			},
+			"runner_id": {
+				Type:        schema.TypeInt,
+				ForceNew:    true,
+				Required:    true,
+				Description: "The ID of a runner to associate to the project",
+			},
+		},
+	}
+})
+
+func resourceGitlabProjectRunnerEnablementCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	projectID := d.Get("project").(string)
+	runnerID := d.Get("runner_id").(int)
+	options := &gitlab.EnableProjectRunnerOptions{
+		RunnerID: runnerID,
+	}
+
+	log.Printf("[DEBUG] create gitlab project runner %v/%v", projectID, runnerID)
+
+	_, _, err := client.Runners.EnableProjectRunner(projectID, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	runnerIDString := strconv.Itoa(runnerID)
+	d.SetId(buildTwoPartID(&projectID, &runnerIDString))
+
+	return resourceGitlabProjectRunnerEnablementRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectRunnerEnablementRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	runnerType := "project_type"
+	projectID, runnerID, err := projectAndRunnerFromID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	options := &gitlab.ListProjectRunnersOptions{
+		ListOptions: gitlab.ListOptions{
+			PerPage: 20,
+			Page:    1,
+		},
+		Type: &runnerType,
+	}
+
+	log.Printf("[DEBUG] read gitlab project runner %s/%v", projectID, runnerID)
+
+	// We're enumerating all the runners for a project to make sure the runner
+	// is indeed associated with the project.
+
+	found := false
+	for {
+		runners, resp, err := client.Runners.ListProjectRunners(projectID, options, gitlab.WithContext(ctx))
+		if err != nil {
+			log.Printf("[DEBUG] failed to read gitlab project runner %s/%v: %s", projectID, runnerID, err)
+			d.SetId("")
+			return diag.FromErr(err)
+		}
+
+		for _, runner := range runners {
+			if runner.ID == runnerID {
+				found = true
+				break
+			}
+		}
+
+		if found || resp.NextPage == 0 {
+			break
+		}
+		options.Page = resp.NextPage
+	}
+
+	if !found {
+		log.Printf("[WARN] removing project runner: %v from state because it no longer exists in gitlab", runnerID)
+		d.SetId("")
+	} else {
+		d.Set("project", projectID)
+		d.Set("runner_id", runnerID)
+	}
+
+	return nil
+}
+
+func projectAndRunnerFromID(id string) (string, int, error) {
+	projectID, runnerIDString, err := parseTwoPartID(id)
+	if err != nil {
+		log.Printf("[WARN] could not get project and runner ids from resource id %v", id)
+		return "", 0, err
+	}
+
+	runnerID, err := strconv.Atoi(runnerIDString)
+	if err != nil {
+		log.Printf("[WARN] could not convert runner id '%s' to integer", runnerIDString)
+	}
+	return projectID, runnerID, err
+
```

Comment:
```suggestion
	return projectID, runnerID, nil
```

Reply:
`err` is not always nil here so I need to return it. But if the flow is unclear I can make it explicitly return `err` inside the if statement

Reply:
Yes, true. I suggest to make it more explicit and also return default type values explicitely for the non-`err` return values.

Reply:
I've made it more explicit